src/charm.py

```diff
@@ -61,49 +50,76 @@
     OATHKEEPER_INFO_INTEGRATION_NAME,
     OPENFGA_INTEGRATION_NAME,
     OPENFGA_STORE_NAME,
-    PEER,
+    PEER_INTEGRATION_NAME,
     PROMETHEUS_SCRAPE_INTEGRATION_NAME,
-    RULES_CONFIGMAP_FILE_NAME,
     TEMPO_TRACING_INTEGRATION_NAME,
-    WORKLOAD_CONTAINER_NAME,
-    WORKLOAD_SERVICE_NAME,
+    WORKLOAD_CONTAINER,
+)
+from exceptions import PebbleError
+from integrations import (
+    HydraIntegration,
+    KratosIntegration,
+    OathkeeperIntegration,
+    OpenFGAIntegration,
+    OpenFGAModelData,
+    PeerData,
+    TracingIntegration,
+)
+from services import PebbleService, WorkloadService
+from utils import (
+    block_when,
+    container_not_connected,
+    integration_not_exists,
+    leader_unit,
+    wait_when,
 )
 
 logger = logging.getLogger(__name__)
 
 
 class IdentityPlatformAdminUIOperatorCharm(CharmBase):
-    """Charm the Identity Platform Admin UI service."""
-
-    def __init__(self, *args):
-        """Charm the service."""
+    def __init__(self, *args: Any):
         super().__init__(*args)
-        self._container = self.unit.get_container(WORKLOAD_CONTAINER_NAME)
 
-        self.hydra_endpoints = HydraEndpointsRequirer(
+        self._container = self.unit.get_container(WORKLOAD_CONTAINER)
+
+        self.peer_data = PeerData(self.model)
+        self._pebble_service = PebbleService(self.unit)
+        self._workload_service = WorkloadService(self.unit)
+
+        self.hydra_endpoints_requirer = HydraEndpointsRequirer(
             self, relation_name=HYDRA_ENDPOINTS_INTEGRATION_NAME
         )
-        self.kratos_info = KratosInfoRequirer(self, relation_name=KRATOS_INFO_INTEGRATION_NAME)
-        self.oathkeeper_info = OathkeeperInfoRequirer(
-            self, relation_name=OATHKEEPER_INFO_INTEGRATION_NAME
-        )
+        self.hydra_integration = HydraIntegration(self.hydra_endpoints_requirer)
 
-        self.ingress = IngressPerAppRequirer(
-            self,
-            relation_name="ingress",
-            port=ADMIN_UI_PORT,
-            strip_prefix=True,
-            redirect_https=False,
+        self.kratos_info_requirer = KratosInfoRequirer(
+            self, relation_name=KRATOS_INFO_INTEGRATION_NAME
         )
+        self.kratos_integration = KratosIntegration(self.kratos_info_requirer)
 
-        self.openfga = OpenFGARequires(self, OPENFGA_STORE_NAME)
+        self.oathkeeper_info_requirer = OathkeeperInfoRequirer(
+            self, relation_name=OATHKEEPER_INFO_INTEGRATION_NAME
+        )
+        self.oathkeeper_integration = OathkeeperIntegration(self.oathkeeper_info_requirer)
 
-        self._admin_ui_cli = AdminUICLI(self._container)
+        self.openfga_requirer = OpenFGARequires(
+            self, store_name=OPENFGA_STORE_NAME, relation_name=OPENFGA_INTEGRATION_NAME
+        )
+        self.openfga_integration = OpenFGAIntegration(self.openfga_requirer)
```

Comment:
is the purpose of the integration to mainly validate payloads and extract variables for pebble?

Reply:
I guess part of? I'd better think of it as a sort of "adapter" of the requirers class. It could the APIs for charm to use to provide needed integration data, validation, and anything related to that specific integration.

Some of them could be very shallow, but some can be deeper. It just depends on how complex that integration could be.

And some integration does not have to do that since it's relatively simple.